Added first version of Readme (#35)
stackabletech/agent: README.adoc

# Stackable Agent

## Purpose
This agent is the component in the Stackable stack which is run locally on all nodes to enact changes that are necessary to deploy and manager the software components that are managed by Stackable.

The agent behaves similar to a  https://github.com/virtual-kubelet/virtual-kubelet[virtual kubelet], but is not implemented using the virtual kubelet framework.

> Instead it is based on the https://github.com/deislabs/krustlet[Krustlet], which is under active development by Microsoft.

Reviewer: ```suggestion
Instead, it is based on the https://github.com/deislabs/krustlet[Krustlet], which is under active development by Microsoft.
```

The agent registers itself as a node with a kube-apiserver will be considered by the Kubernetes scheduler for workloads (pods).
To avoid _normal_ Kubernetes pods being scheduled on the Stackable agent (it would not know what to do with these) the agent assigns the following taints to its node object:

|===
|Taint |Type|Value

|kubernetes.io/arch
|NoSchedule
|stackable-linux

|kubernetes.io/arch
|NoExecute
|stackable-linux
|===

These taints _suggest_ to the Kubernetes scheduler that only pods with matching tolerations should be scheduled on this node.

> ## Installation

Reviewer: ```suggestion
== Installation
```

> ### Build from source

Reviewer: ```suggestion
=== Build from source
```

Building from source is fairly straightforward if you have the rust toolchain installed.
If you need to install this, generally the recommended way is to use https://rustup.rs/[rustup].

After rust is installed simply run

    cargo build

To create the binary file in _target/debug_.

> ### Download binary

Reviewer: ```suggestion
=== Download binary
```

We do not at this time provide pre-compiled binaries, as we are still in the process of setting up the build jobs to create these.
This readme will be updated as soon as binary downloads are available!

> ### OS Packages

Reviewer: ```suggestion
=== OS Packages
```

> We do not at this time provide os packages, that is due to change in the near future and this readme will be updated accordingly!

Reviewer: ```suggestion
We do not at this time provide OS packages, that is due to change in the near future and this readme will be updated accordingly!
```

> ## Configuration

Reviewer: ```suggestion
== Configuration
```

> ### Command Line Parameters

Reviewer: ```suggestion
=== Command Line Parameters
```

> WARNING: Due to https://github.com/stackabletech/agent/issues/34[#34] it is not currently possible to specify these parameters directly on the commandline, please use a config file instead.

Reviewer: ```suggestion
WARNING: Due to https://github.com/stackabletech/agent/issues/34[issue #34] it is not currently possible to specify these parameters directly on the commandline, please use a config file instead.
```

The agent accepts the following command line parameters:

include::config/commandline_args.adoc[]

> ### Config File

Reviewer: ```suggestion
=== Config File
```

> In addition to directly specifying them on the command line, the agent allows specifying a config file via the environment variable _CONFIG_FILE_. Values specified in the file will have to adhere to the format _--parameter=value_.

Reviewer: ```suggestion
In addition to directly specifying them on the command line, the agent allows specifying a config file via the environment variable `CONFIG_FILE`. Values specified in the file will have to adhere to the format `--parameter=value`.
```

> This file can contain all command line parameters as well and will be parsed before the actual command line.

Reviewer: ```suggestion
This file can contain all command line parameters and will be parsed before the actual command line.
```

> For parameters that are present in the file and on the command line, the command line will take precedence, unless it is a parameter that can be specified multiple times, in which case paramaters from both, file and commandline, will be used.

Reviewer: ```suggestion
For parameters that are present in the file and on the command line, the command line will take precedence, unless it is a parameter that can be specified multiple times, in which case parameters from both, file and commandline, will be merged.
```

.Example config file
    --package-directory=/opt/stackable/agent/work/packages
    --config-directory=/etc/stackable/agent
    --server-cert-file=/etc/stackable/agent/secure/cert.crt
    --server-key-file=/etc/stackable/agent/secure/key.key

> ### Kubernetes Config

Reviewer: ```suggestion
=== Kubernetes Config
```

The agent uses the default way of looking for a kube-apiserver, so if your system is already set up to connect to Kubernetes with kubectl you should be good to go right of the bat.

> The default location for the Kubernetes client config is _~/.kube/config_, if you want to change this location you can override this via the _KUBECONFIG_ environment variable.

Reviewer: ```suggestion
The default location for the Kubernetes client config is `~/.kube/config`, if you want to change this location you can override this via the `KUBECONFIG` environment variable.
```

    export KUBECONFIG=/etc/stackable/agent/kubeconfig

> ### Certificates

Reviewer: ```suggestion
=== Certificates
```

The agent requires a keypair and signed certificate to start a webserver which can be used to handle callbacks.
If these are not specified on the commandline, the agent will create a keypair, upload a certificate signing request to Kubernetes and wait for the certificate before continuing.
These steps require a certificate manager to be set up and running in your Kubernetes cluster, which may or may not be the case.

You can also manually create these files and specify them on the command line.
The following example shows how to create these files using https://github.com/OpenVPN/easy-rsa[easy-rsa], but this can be done in any number of different ways as well.

    ./easyrsa init-pki
    ./easyrsa build-ca
    ./easyrsa gen-req krustlet1
    ./easyrsa import-req pki/reqs/krustlet1.req krustlet1-req
    ./easyrsa sign-req serverClient krustlet1-req
    # Convert key to pksc8 format
    openssl pkcs8 -topk8 -inform PEM -outform PEM -nocrypt -in pki/private/krustlet1.key -out pkcs8.key

> ## Contributing

Reviewer: ```suggestion
== Contributing
```

= Stackable Agent

== Purpose

> The purpose of the Stackable architecture, if I had to put it in one sentence is to "run Kubernetes workloads without containers".

Reviewer: I think I disagree with that as it's our plan to also run containers and this readme should be about the agent and not the whole architecture. At least the beginning.

Why not write something along these lines: "The Stackable Agent is an alternative to the Kubernetes Kubelet that executes Pods not in containers but using systemd as its backend. It is implemented in Rust as a Krustlet provider." (+ link to krustlet)

Author: I've tried again, if this still doesn't fit the bill I'd suggest that we get together with a gdoc and collaborate directly on a final version.

Instead components are installed locally on the host filesystem and executed via systemd units.

This allows using the fabulous orchestration options offered by Kubernetes to describe the necessary infrastructure while leaving aside the extra overhead and complexity that are added by running services in containers.
Additionally, as we are already using a Kubernetes cluster, this will fairly easily allow defining hybrid scenarios where some services run on pure Kubernetes and others are orchestrated on bare-metal.

The agent in principle behaves similar to a https://github.com/virtual-kubelet/virtual-kubelet[virtual kubelet], but is not implemented using the virtual kubelet framework.
Instead it is based on the https://github.com/deislabs/krustlet[Krustlet], which is under active development by Microsoft.

More specifically the functionality of this agent closely resembles that of https://github.com/virtual-kubelet/systemk[systemk].
The reason why both projects currently exist is simply that systemk was not around when we started Stackable and so we had to go and build our own.

> We are currently evaluating whether or not we should instead use systemk and contribute to that project instead of building our own agent under issue

Reviewer: link + sentence end missing
